[workers] Write r2 finetune tutorial (#10754)
cloudflare/cloudflare-docs: content/workers/tutorials/create-finetuned-chatgpt-ai-models-with-r2/index.md

In order to use this feature, you will do the following tasks:

1. Upload a fine-tune document to R2
2. Read the R2 file and upload it to OpenAI
3. Create a new fine-tuned model based on the document

![Demo](/images/workers/tutorials/finetune/finetune-example.png)

If you'd like to see the completed code for this application, you can find it [on GitHub](https://github.com/kristianfreeman/openai-finetune-r2-example).

## Prerequisites

Before you start, make sure you have:

- A Cloudflare account. If you do not have one, [sign up](https://dash.cloudflare.com/sign-up/workers-and-pages) before continuing.
- An OpenAI API key.
- A fine-tune document, structured as [JSON Lines](https://jsonlines.org/). You can use the [example document](https://github.com/kristianfreeman/openai-finetune-r2-example/blob/16ca53ca9c8589834abe317487eeedb8a24c7643/example_data.jsonl) in the source code.

> 

Author: sure we can add it! do we have an existing one we use?

> 

Reviewer: Look at the partial code on https://developers.cloudflare.com/workers/tutorials/deploy-a-realtime-chat-app/

> 

Reviewer: This should be on all tutorials

		file,
		purpose: "fine-tune",
	})

	return uploadedFile
}

app.get('/files', async c => {
	const fileQueryParam = c.req.query("file")
	if (!fileQueryParam) return c.text("Missing file query param", 400)

	const file = await c.env.ASSETS.get(fileQueryParam)
	if (!file) return c.text("Couldn't find file", 400)

	const uploadedFile = await createFile(c, file)
	return c.json(uploadedFile)
})
```

### 5. Creating fine-tuned models

This section includes the `GET /models` route and the `createModel` function. The route handles incoming requests for creating a new fine-tuned model. The function `createModel` takes care of specifying the details and initiating the fine-tuning process with OpenAI.

```javascript
---
filename: src/worker.js
---
const createModel = async (c: Context, fileId: string) => {
	const openai: OpenAI = c.get("openai")

	const body = {
		training_file: fileId,
		model: "gpt-3.5-turbo",
	}

	return openai.fineTuning.jobs.create(body)
}

app.get('/models', async c => {
	const fileId = c.req.query("file_id")
	if (!fileId) return c.text("Missing file ID query param", 400)

	const model = await createModel(c, fileId)
	return c.json(model)
})
```

> ### 6. Listing all fine-tune jobs

Reviewer: ```suggestion
### 6. List all fine-tune jobs
```

This section describes the `GET /jobs` route and the corresponding `getJobs` function. The route provides an interface for retrieving a list of all fine-tuning jobs. The function interacts with OpenAI's API to fetch and return this information.

```javascript
---
filename: src/worker.js
---
app.get('/jobs', async c => {
	const jobs = await getJobs(c)
	return c.json(jobs)
})

const getJobs = async (c: Context) => {
	const openai: OpenAI = c.get("openai")
	const resp = await openai.fineTuning.jobs.list()
	return resp.data
}
```

### 7. Deploy your application

> Once you have created your Worker application and added the required functions, deploy the application. 

Reviewer: ```suggestion
After you have created your Worker application and added the required functions, deploy the application. 
```

> Before you deploy, you must set the `OPENAI_API_KEY` [secret](/workers/configuration/secrets/) for your application. You can do this by running the [`npx wrangler secret put`](/workers/wrangler/commands/#put-3) command:

Reviewer: ```suggestion
Before you deploy, you must set the `OPENAI_API_KEY` [secret](/workers/configuration/secrets/) for your application. Do this by running the [`npx wrangler secret put`](/workers/wrangler/commands/#put-3) command:
```

```sh
$ npx wrangler secret put OPENAI_API_KEY
```

To deploy your Worker application to the Cloudflare global network:

1. Make sure you are in your Worker project's directory, then run the [`npx wrangler deploy`](/workers/wrangler/commands/#deploy) command:

```sh
$ npx wrangler deploy
```

2. Wrangler will package and upload your code.

3. After your application is deployed, Wrangler will provide you with your Worker's URL.

> ### 8. Viewing the fine-tune job status, and using the model

Reviewer: ```suggestion
### 8. View the fine-tune job status and use the model
```

To use your application, create a new fine-tune job by making a request to the `/files` with a `file` query param matching the filename you uploaded earlier:

```sh

> $ curl yourworker/files?file=finetune.jsonl

Reviewer: Can we be more clear on what 'yourworker' is? Is it the deployed URL? Directory name?

Author: updating to `https://your-worker-url.com`

```

When the file is uploaded, issue another request to `/models`, passing the `file_id` query parameter. This should match the `file_id` returned as JSON from the `/files` route:

```sh
$ curl yourworker/models?file_id=file-abc123
```

Finally, visit `/jobs` to see the status of your fine-tune jobs in OpenAI. Once the fine-tune job has completed, you can see the `fine_tuned_model` value, indicating a fine-tuned model has been created.

![Jobs](/images/workers/tutorials/finetune/finetune-jobs.png)

> You can visit the [OpenAI Playground](https://platform.openai.com/playground) in order to use your fine-tune model. Select it in the "Model" section on the right sidebar of the interface.

Reviewer: ```suggestion
Visit the [OpenAI Playground](https://platform.openai.com/playground) in order to use your fine-tune model. Select your fine-tune model in the **Model** section on the right sidebar of the interface.
```
Define 'it'.

![Demo](/images/workers/tutorials/finetune/finetune-example.png)

> You can also use it in any API requests you make to OpenAI's "chat completions" endpoints. For instance, in the below code example:

Reviewer: ```suggestion
Use it in any API requests you make to OpenAI's chat completions endpoints. For instance, in the below code example:
```
No "" in tech docs. Should this be monospaced instead? Looking at OpenAI docs, they left it like I did.
